### PR TITLE
Test and diagnostics improvements

### PR DESCRIPTION
**Pull request by keszybz**

### CODE REVIEW

File: src/basic/mkdir.h [hunk 1]
```diff
@@ -25,6 +25,7 @@
 
 typedef enum MkdirFlags {
         MKDIR_FOLLOW_SYMLINK = 1 << 0,
+        MKDIR_WARN_MODE      = 1 << 1,
```
Comment (poettering) [inline src/basic/mkdir.h]: this does more than "warn", right? it also refuses, hence maybe call thus MKDIR_REFUSE_AND_WARN_MODE or so?

Comment (poettering) [reply src/basic/mkdir.h]: Or MKDIR_PEDANTIC? MKDIR_FASCIST? MKDIR_THROW_TANTRUM? ;-)

Comment (poettering) [reply src/basic/mkdir.h]: also, it checks more than just the mode. hence maybe MKDIR_COMPLAIN_STRICT_PERMS?

Comment (poettering) [reply src/basic/mkdir.h]: or maybe split this up: MKDIR_WARN as one flag, and MKDIR_STRICT_PERMISSIONS as another one? And then doe the usual `log_full(flags & MKDIR_WARN ? LOG_ERR : LOG_DEBUG, …` dance? I figure that would be the most systematic approach?

Comment (keszybz) [reply src/basic/mkdir.h]: No, this flag only influences logging. (At least that was the intent. But looking at the code, it seems correct.)

I'll rework the patch to use `log_full` though.

**Comment by keszybz** [comment]
Update to use log_full.

File: src/basic/mkdir.c [hunk 2]
```diff
@@ -61,13 +62,32 @@ int mkdir_safe_internal(const char *path, mode_t mode, uid_t uid, gid_t gid, Mkd
                         return -errno;
         }
 
+        if (!S_ISDIR(st.st_mode)) {
+                log_full(flags & MKDIR_WARN_MODE ? LOG_WARNING : LOG_INFO,
```
Comment (poettering) [inline src/basic/mkdir.c]: LOG_INFO → LOG_DEBUG

File: src/basic/mkdir.c [hunk 3]
```diff
@@ -61,13 +62,32 @@ int mkdir_safe_internal(const char *path, mode_t mode, uid_t uid, gid_t gid, Mkd
...
+                return -ENOTDIR;
+        }
         if ((st.st_mode & 0007) > (mode & 0007) ||
             (st.st_mode & 0070) > (mode & 0070) ||
-            (st.st_mode & 0700) > (mode & 0700) ||
-            (uid != UID_INVALID && st.st_uid != uid) ||
-            (gid != GID_INVALID && st.st_gid != gid) ||
-            !S_ISDIR(st.st_mode))
+            (st.st_mode & 0700) > (mode & 0700)) {
+                log_full(flags & MKDIR_WARN_MODE ? LOG_WARNING : LOG_INFO,
```
Comment (poettering) [inline src/basic/mkdir.c]: dito

File: src/basic/mkdir.c [hunk 4]
```diff
@@ -61,13 +62,32 @@ int mkdir_safe_internal(const char *path, mode_t mode, uid_t uid, gid_t gid, Mkd
...
+        }
+        if ((uid != UID_INVALID && st.st_uid != uid) ||
+            (gid != GID_INVALID && st.st_gid != gid)) {
+                char u[DECIMAL_STR_MAX(uid_t)] = "-", g[DECIMAL_STR_MAX(gid_t)] = "-";
+
+                if (uid != UID_INVALID)
+                        xsprintf(u, UID_FMT, uid);
+                if (gid != UID_INVALID)
+                        xsprintf(g, GID_FMT, gid);
+                log_full(flags & MKDIR_WARN_MODE ? LOG_WARNING : LOG_INFO,
```
Comment (poettering) [inline src/basic/mkdir.c]: dito

File: src/basic/mkdir.c [hunk 5]
```diff
@@ -61,13 +62,32 @@ int mkdir_safe_internal(const char *path, mode_t mode, uid_t uid, gid_t gid, Mkd
...
+                         "Directory \"%s\" already exists, but is owned by %u:%u (%s:%s was requested), refusing.",
```
Comment (poettering) [inline src/basic/mkdir.c]: UID_FMT ":" GID_FMT …

**Comment by keszybz** [comment]
Pfff, sorry for that. Updated.

**Pull request merged**